Services/PredefinedValuesProductAttributeService.cs

```diff
@@ -0,0 +1,47 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using OrchardCore.Commerce.Abstractions;
+using OrchardCore.Commerce.Models;
+using OrchardCore.Commerce.Settings;
+using OrchardCore.ContentManagement;
+
+namespace OrchardCore.Commerce.Services
+{
+    public class PredefinedValuesProductAttributeService : IPredefinedValuesProductAttributeService
+    {
+        private readonly IProductAttributeService _productAttributeService;
+
+        public PredefinedValuesProductAttributeService(IProductAttributeService productAttributeService)
+        {
+            _productAttributeService = productAttributeService;
+        }
+
+        public IEnumerable<IEnumerable<string>> GetProductAttributesPredefinedValues(ContentItem product)
+        {
+            return _productAttributeService.GetProductAttributeFields(product)
+                .Where(x => x.Settings is TextProductAttributeFieldSettings textSettings && textSettings.RestrictToPredefinedValues)
```

Comment:
Ah, so for this I thought you could discover what attribute types implement some `IPredefinedValuesProductAttribute` interface, instead of hard-coding for the text attribute.

Reply:
Done